starlark/hashtable.go

```diff
@@ -371,3 +364,186 @@ func softHashString(s string) uint32 {
 	}
 	return h
 }
+
+// An OrderedStringDict is a mapping from names to values with support
+// for fast indexing. Keys are stored in order of unique insertion.
+// Keys are fast to add and access but slow to delete.
+// It is not a true starlark.Value.
+type OrderedStringDict struct {
+	// Implementation based on the above hashtable.
+	table   []osdBucket     // len is zero or a power of two
+	bucket0 [bucketSize]int //*osdEntry // inline allocation for small maps.
+	entries []osdEntry      // sorted list of entries
+}
+
+type osdBucket []int //*osdEntry
+
+type osdEntry struct {
+	hash  uint32
+	key   string
+	value Value
+}
+
+func (d *OrderedStringDict) init(size int) {
+	if size < 0 {
+		panic("size < 0")
+	}
+	nb := 1
+	for overloaded(size, nb) {
+		nb = nb << 1
+	}
+	if nb < 2 {
+		d.table = []osdBucket{d.bucket0[:0]}
+	} else {
+		d.table = make([]osdBucket, nb)
+		for i := range d.table {
+			d.table[i] = make(osdBucket, 0, bucketSize)
+		}
+	}
+	d.entries = make([]osdEntry, 0, size)
+}
+
+func (d *OrderedStringDict) rehash() {
+	// TODO: shrink?
+	for i, l := 0, len(d.table); i < l; i++ {
+		d.table[i] = d.table[i][:0]
+	}
+	oldEntries := d.entries
+	d.entries = d.entries[:0]
+	for i, l := 0, len(oldEntries); i < l; i++ {
+		e := oldEntries[i]
+		d.insert(e.hash, e.key, e.value)
+	}
+}
+
+func (d *OrderedStringDict) grow() {
+	// Double the number of buckets and rehash.
+	newTable := make([]osdBucket, len(d.table)<<1)
+	for i, l := 0, len(d.table); i < l; i++ {
+		// Reuse bucket if below bucketSize.
+		p := d.table[i]
+		if cap(p) <= bucketSize {
+			newTable[i] = p[:0]
+		} else {
+			newTable[i] = make(osdBucket, 0, bucketSize)
+		}
+	}
+	for i, l := len(d.table), len(newTable); i < l; i++ {
+		newTable[i] = make(osdBucket, 0, bucketSize)
+	}
+	d.table = newTable
+	oldEntries := d.entries
+	d.entries = make([]osdEntry, 0, len(d.entries)<<1)
+	for i, l := 0, len(oldEntries); i < l; i++ {
+		e := oldEntries[i]
+		d.insert(e.hash, e.key, e.value)
+	}
+}
+
+func (d *OrderedStringDict) insert(h uint32, k string, v Value) {
+	if d.table == nil {
+		d.init(1)
+	}
+
+	// Does the number of elements exceed the buckets' load factor?
+	for overloaded(len(d.entries), len(d.table)) {
+		d.grow()
+	}
+
+	n := h & (uint32(len(d.table) - 1))
+	p := d.table[n]
+	for i, l := 0, len(p); i < l; i++ {
+		e := &d.entries[p[i]]
+		if h == e.hash && k == e.key {
+			e.value = v
+			return
+		}
+	}
+
+	// Append value to entries, linking the bucket to the entries list.
+	d.entries = append(d.entries, osdEntry{
+		hash:  h,
+		key:   k,
+		value: v,
+	})
+	d.table[n] = append(p, len(d.entries)-1)
+}
+
+func OrderStringDict(d StringDict) OrderedStringDict {
```

Comment:
The doc comment should state that the resulting value's fields are ordered by name.